molecule/src/structure_checker2.cpp

```diff
@@ -0,0 +1,651 @@
+{
+    if (mol.tgroups.getTGroupCount() > 0)
+    {
+        result.message(StructureChecker2::CheckMessageCode::CHECK_MSG_TGROUP);
+    }
+}
+static void check_chirality(Molecule& mol, const std::unordered_set<int>& selected_atoms, const std::unordered_set<int>& selected_bonds,
+                            StructureChecker2::CheckResult& result)
+{
+    // not impl
+    result.message(StructureChecker2::CheckMessageCode::CHECK_MSG_CHIRALITY);
+}
+static void check_chiral_flag(Molecule& mol, const std::unordered_set<int>& selected_atoms, const std::unordered_set<int>& selected_bonds,
+                              StructureChecker2::CheckResult& result)
+{
+    if (mol.getChiralFlag() > 0 && mol.stereocenters.size() == 0)
+    {
+        result.message(StructureChecker2::CheckMessageCode::CHECK_MSG_CHIRAL_FLAG);
+    }
+}
+static void check_3d_coord(Molecule& mol, const std::unordered_set<int>& selected_atoms, const std::unordered_set<int>& selected_bonds,
+                           StructureChecker2::CheckResult& result)
+{
+    FILTER_ATOMS(StructureChecker2::CheckMessageCode::CHECK_MSG_3D_COORD, [](Molecule& mol, int idx) { return fabs(mol.getAtomXyz(idx).z) > 0.001; });
+}
+static void check_charge(Molecule& mol, const std::unordered_set<int>& selected_atoms, const std::unordered_set<int>& selected_bonds,
+                         StructureChecker2::CheckResult& result)
+{
+    // not impl
+    result.message(StructureChecker2::CheckMessageCode::CHECK_MSG_CHARGE_NOT_IMPL);
+}
+static void check_salt(Molecule& mol, const std::unordered_set<int>& selected_atoms, const std::unordered_set<int>& selected_bonds,
+                       StructureChecker2::CheckResult& result)
+{
+    // not impl
+    result.message(StructureChecker2::CheckMessageCode::CHECK_MSG_SALT_NOT_IMPL);
+}
+static void check_ambigous_h(Molecule& mol, const std::unordered_set<int>& selected_atoms, const std::unordered_set<int>& selected_bonds,
+                             StructureChecker2::CheckResult& result)
+{
+    if (mol.isQueryMolecule())
+    {
+        result.message(StructureChecker2::CheckMessageCode::CHECK_MSG_AMBIGUOUS_H_NOT_CHECKED_QUERY);
+    }
+    else
+    {
+        FILTER_ATOMS_DEFAULT(StructureChecker2::CheckMessageCode::CHECK_MSG_AMBIGUOUS_H, [](Molecule& mol, int idx) {
+            return mol.asMolecule().getImplicitH_NoThrow(idx, -1) == -1 && mol.getAtomAromaticity(idx) == ATOM_AROMATIC;
+        });
+    }
+}
+static void check_coord(Molecule& mol, const std::unordered_set<int>& selected_atoms, const std::unordered_set<int>& selected_bonds,
+                        StructureChecker2::CheckResult& result)
+{
+    if (mol.vertexCount() > 1 && !BaseMolecule::hasCoord(mol))
+    {
+        result.message(StructureChecker2::CheckMessageCode::CHECK_MSG_ZERO_COORD);
+    }
+}
+static void check_v3000(int mol, StructureChecker2::CheckResult& result)
+{
+    const char* f = indigoMolfile(mol);
```

Comment:
there is a dependency violation here - core libraries (common, graph, molecule, layout, reaction) should not depend on API. This works for you now only because you don't build Bingo (which depends on molecule, but not on API).

And also ID only makes sense in Indigo session, not even in IndigoObject - objects could potentially be even serialized/deserialized between sessions and they should not now their ID.

So I suggest passsing here `BaseMolecule` instead of id.

Reply:
Regarding to dependency violations:
![image](https://user-images.githubusercontent.com/76469602/104799488-93451100-57e0-11eb-9ec3-3b15ef5dd117.png)
"Check" functionality is spreaded across the code.
molfile_loader performs some checks itself instead of loading of any syntactically correct file assuming that the result may/should be checked later by the "checker" which concentrates all checking functionality.

Therefore you can't check molecule containing errors, i.e. exactly for what checking is necessary: you just can't load it without not-evident dancing around "ignore errors" options.

Also note that "ex" is added by me:
Originally it throws anonymous exception and fails silently:
       catch (Error&)
        {
            if (!options.ignore_errors)
                throw;
        }

Reply:
Regarding to dependency violation:

I see that API implements features - see below.
Core should call API for the feature? - this is a dependency violation
API is not an interface to expose core to users but it implements features itself? How to use the features in core programming? Duplicate them?

In the particular case below, why we write the "checker" if we don't use it for checking?
Checking facilities appear in many unpredictable places (see above and below) including Python REST service. 


CEXPORT int indigoCheck3DStereo(int item)
{
    INDIGO_BEGIN
    {
        IndigoObject& obj = self.getObject(item);

        if (IndigoBaseMolecule::is(obj))
        {
            BaseMolecule& bmol = obj.getBaseMolecule();

            bool stereo_3d = true;

            if ((bmol.stereocenters.size() > 0) && BaseMolecule::hasZCoord(bmol))
            {
                for (auto i : bmol.vertices())
                {
                    if (!bmol.stereocenters.exists(i))
                    {
                        const Vertex& vertex = bmol.getVertex(i);

                        for (auto j = vertex.neiBegin(); j != vertex.neiEnd(); j = vertex.neiNext(j))
                            if (bmol.getBondDirection2(i, vertex.neiVertex(j)) > 0)
                                stereo_3d = false;
                    }
                }
            }
            else
                stereo_3d = false;

            if (stereo_3d)
                return 1;
        }
        return 0;
    }
    INDIGO_END(-1);
}

Reply:
Indigo/api/src/indigo_stereo.cpp

500 rows of code. 
Including private funcs.
This is not API, this is core functionality

Should core use it? 
Should core call it using handles?
CEXPORT int indigoValidateChirality(**int handle**)